thrift/thrift0.13/src/main/java/com/linecorp/armeria/internal/server/thrift/ThriftMetadataAccess.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.server.thrift;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.Properties;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.thrift.TBase;
+import org.apache.thrift.TFieldIdEnum;
+import org.apache.thrift.meta_data.FieldMetaData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+final class ThriftMetadataAccess {
+
+    private static final Logger logger = LoggerFactory.getLogger(ThriftMetadataAccess.class);
+
+    private static boolean preInitializeThriftClass;
+    private static final Pattern preInitializeTargetPattern =
+            Pattern.compile("^armeria-thrift0\\.(\\d+)\\..*$");
+
+    static {
+        try {
+            final Enumeration<URL> versionPropertiesUrls =
+                    ThriftMetadataAccess.class.getClassLoader().getResources(
+                            "META-INF/com.linecorp.armeria.versions.properties");
```

Comment:
Two questions:

- Can't we just do `Version.getAll(...).keySet()` to get the list of the artifact IDs?
- Can't we just always pre-initialize the classes?
- What do you think about adding an empty resource file like `src/main/resources/com/linecorp/armeria/internal/common/thrift/requires_struct_preinit` to all thrift modules older than 0.15? We could enable the workaround only if that resource file exists.

Reply:
Given the cost of failure to look for a resource file might be a little bit higher, we could always add a properties file:

If pre-init is required:

```
structPreinitRequired=true
```

If pre-init is not required:

```
<empty file>
```

Reply:
> Can't we just do Version.getAll(...).keySet() to get the list of the artifact IDs?

Didn't know this class existed 😅   It's probably better to use this.

> Can't we just always pre-initialize the classes?

I think I prefer making an explicit branch statement for these type of changes:
1) If we decide to deprecate thrift <= 0.14 modules, we can remove the related logic safely
2) The reason for pre-initialization is embedded in the code

I do agree that the logic doesn't really affect performance and am open to just preinitializing always if others feel this is simpler.

> What do you think about adding an empty resource file
> we could always add a properties file

I don't even think this has to be a resource, but it can also be a class file that has a different static value for each thrift module. I didn't want to further complicate the thrift file copying logic, but I think this is also an option.

I think I prefer just using `Versions.all()` like you mentioned, but let me know if you feel differently.

Reply:
> this doesn't work if multiple armeria-thrift artifacts are added.

Doesn't a user have a bigger problem if their classpath has multiple `armeria-thrift` artifacts? We could leave a warning and fall back to safe mode (always preinit)?

Reply:
> I didn't think it was possible to detect if there are multiple classes with the same package/name/module.

You can actually:

```java
Enumeration<URL> urls = getClass().getClassLoader().getResources("com/foo/bar.txt"); // or .class
while (urls.hasMoreElements()) {
    URL url = urls.nextElement();
    InputStream inputStream = url.openStream();
    // do something with the input stream
}
```

Otherwise, how could `Versions.all()` fetch the `com.linecorp.armeria.versions.properties` files from all artifacts?

Reply:
By the way, I meant putting some `.properties` file rather than looking for `.class` files. I think having a separate resource is still better than extracting version number from artifact ID because it doesn't require any string matching, which could be broken at some point.

Reply:
>  I think having a separate resource is still better than extracting version number from artifact ID because it doesn't require any string matching, which could be broken at some point.

I see 😄    Understood.